resources/idlharness.js

```diff
@@ -3394,16 +3407,66 @@ function idl_test(srcs, deps, idl_setup_func) {
     }, 'idl_test setup');
 }
 
-/**
- * fetch_spec is a shorthand for a Promise that fetches the spec's content.
- */
-function fetch_spec(spec) {
-    var url = '/interfaces/' + spec + '.idl';
+// TODO: it would be nice to support `idl_array.add_objects`
+function idl_test_shadowrealm(srcs, deps) {
+    const script_urls = [
+        "/resources/testharness.js",
+        "/resources/WebIDLParser.js",
+        "/resources/idlharness.js",
+    ];
+    promise_test(t => {
+        const realm = new ShadowRealm();
+        realm.evaluate("globalThis.self = globalThis; undefined");
+
+        return Promise.all(script_urls.map(url => fetch_text(url).then(s => realm.evaluate(s))))
+        .then(_ => Promise.all(srcs.concat(deps).map(fetch_spec)))
+        .then(specs => {
+            const idls = JSON.stringify(specs.map(i => i.idl));
+            const code = `
+                const idls = JSON.parse(${JSON.stringify(idls)});
+                let results;
+                add_completion_callback(function (tests, harness_status, asserts_run) {
+                  results = tests;
+                });
+
+                // TODO: maybe the indexed iteration as in idl_test is better.
+                const [srcs, deps] = [idls.slice(0, ${srcs.length}), idls.slice(${srcs.length})];
+                test(() => {
+                  var idl_array = new IdlArray();
+                  for (const src of srcs) {
+                    idl_array.add_idls(src);
+                  }
+                  for (const dep of deps) {
+                    idl_array.add_dependency_idls(dep);
+                  }
+                  idl_array.test();
+                }, "inner setup");
```

Comment:
Can we avoid this extra test and have just `idl_test setup`?

Reply:
Without it, testharness.js will think we're done after the first idlharness-generated subtest. I'm not actually quite sure why, but I recall that code being fairly brittle.